replication.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package simplex
+
+import (
+	"bytes"
+
+	"go.uber.org/zap"
+)
+
+type Request struct {
+	FinalizationCertificateRequest *FinalizationCertificateRequest
+	LatestRoundRequest             *LatestRoundRequest
+}
+
+type Response struct {
+	FinalizationCertificateResponse *FinalizationCertificateResponse
+	LatestRoundResponse             *LatestRoundResponse
+}
+
+// request a finalization certificate for the given sequence number
+type FinalizationCertificateRequest struct {
+	Seq uint64
+}
+
+type FinalizationCertificateResponse struct {
+	FCert FinalizationCertificate
+	Block Block
+}
+
+type LatestRoundRequest struct {
+}
+
+// if block has not been proposed yet, we should return the last round's block, notarization, and fCert
+type LatestRoundResponse struct {
+	Block        Block
+	Notarization *Notarization
+	FCert        *FinalizationCertificate
+}
+
+func (e *Epoch) HandleRequest(req Request) *Response {
+	// TODO: should I update requests to be async? and have the epoch respond with e.Comm.Send(msg, node)
+	response := &Response{}
+
+	if req.FinalizationCertificateRequest != nil {
+		response.FinalizationCertificateResponse = e.handleFinalizationCertificateRequest(req.FinalizationCertificateRequest)
+	}
+	if req.LatestRoundRequest != nil {
+		response.LatestRoundResponse = e.handleLatestBlockRequest()
+	}
+
+	return response
+}
+
+func (e *Epoch) handleFinalizationCertificateRequest(req *FinalizationCertificateRequest) *FinalizationCertificateResponse {
+	block, fCert, exists := e.Storage.Retrieve(req.Seq)
+	if !exists {
+		return nil
+	}
+	return &FinalizationCertificateResponse{
+		FCert: fCert,
+		Block: block,
+	}
+}
+
+func (e *Epoch) handleLatestBlockRequest() *LatestRoundResponse {
+	round, ok := e.rounds[e.round]
+	// this means we have not proposed a block yet for this round
+	if !ok {
+		if e.round == 0 {
+			return &LatestRoundResponse{}
+		}
+		// get the previous round from memory or storage
+		round, ok := e.rounds[e.round-1]
+		if ok {
+			return &LatestRoundResponse{
+				Block:        round.block,
+				Notarization: round.notarization,
+				FCert:        round.fCert,
+			}
+		}
+
+		// get latest block
+		block, fCert, err := RetrieveLastFromStorage(e.Storage)
+		if err != nil {
+			return &LatestRoundResponse{}
+		}
+		return &LatestRoundResponse{
+			Block:        block,
+			FCert:        &fCert,
+			Notarization: nil,
+		}
+	}
+
+	latest := &LatestRoundResponse{
```

Comment:
It's not possible to have the same `e.round` both notarized and finalized at the same time.

If we have notarized the round, we'd move to `e.round+1` immediately and may or may not finalize it. 

Therefore, `e.round` always represents the current round we're waiting to notarize at the moment.

I think that the simplest thing to do, is to just always grab the latest finalization we have, and the latest notarization we have, regardless of the round we're at.